### PR TITLE
adapter: add mz_materialization_lag view

### LINKED ISSUE #21162: Query for tracking materialization lag
Now that we have both compute and source frontiers exposed in SQL, we can write queries that return the materialization lag of each compute collection. We anticipate that such queries will be useful for our internal metrics, but also to users for monitoring the health of their dataflows.

There are two types of lag we are potentially interested in:

* The end-to-end delay between data becoming available at the sources and a compute collection having processed the data.
* The compute-internal delay between data becoming available to a compute cluster and a compute collection on that cluster having processed the data.

In both cases, the formula to calculate the delay is (assuming that frontiers are epoch-ms times): `min(<input frontiers>) - <compute collection frontier>`

What differs is the selection of `<input frontiers>`. For end-to-end lag we want the frontiers of all sources and tables that feed the compute collection. For compute-internal lag we want to break the dependency chain at any persist collection (which includes MVs, in addition to sources and tables).

### PR DESCRIPTION
**Pull request by teskje**
This PR adds a view that reports the local (from direct inputs) and global (from root inputs) materialization lag.

The `slowest_*` columns were added on @SangJunBak's request as they might be useful in finding the first lagging object in a dependency graph.

### Motivation

  * This PR adds a known-desirable feature.

Closes #21162 

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add a new built-in view, `mz_internal.mz_materialization_lag` that exposes the amount of time outputs from materialized views, indexes, and sinks lag behind their inputs.

### CODE REVIEW

**Comment by SangJunBak** [comment]
This fits what I need@

**Comment by SangJunBak** [comment]
Would we also be able to create an index on this view with this PR or another? When running against 400 stacked views(linked list) that each have an index, the query takes roughly 500ms to 2s. Either that or index mz_frontiers, whatever is causing the bottleneck!

Explain plan:
```
Optimized Plan
"Explained Query:
  Return
    Return
      Project (#0, #9, #10, #3, #6)
        Map ((#1) IS NULL, case when #8 then 00:00:00 else case when (#4) IS NULL then null else greatest((#4 - #1), 00:00:00) end end, case when #8 then 00:00:00 else case when (#7) IS NULL then null else greatest((#7 - #1), 00:00:00) end end)
          Join on=(eq(#0, #2, #5)) type=delta
            ArrangeBy keys=[[#0]]
              Project (#0, #3)
                Map (tots((text_to_double(mz_timestamp_to_text(#2)) / 1000)))
                  Join on=(#0 = #1) type=differential
                    ArrangeBy keys=[[#0]]
                      Get l0
                    Get l5
            ArrangeBy keys=[[#0]]
              Project (#0, #1, #3)
                TopK group_by=[#0] order_by=[#2 asc nulls_last] limit=1
                  Project (#0, #1, #3, #4)
                    Map (tots((text_to_double(mz_timestamp_to_text(#3)) / 1000)))
                      Join on=(#1 = #2) type=differential
                        ArrangeBy keys=[[#1]]
                          Get l2
                        Get l5
            ArrangeBy keys=[[#0]]
              Project (#0, #1, #3)
                TopK group_by=[#0] order_by=[#2 asc nulls_last] limit=1
                  Project (#0, #1, #4, #5)
                    Map (tots((text_to_double(mz_timestamp_to_text(#4)) / 1000)))
                      Join on=(eq(#1, #2, #3)) type=delta
                        ArrangeBy keys=[[#1]]
                          Get l3
                        ArrangeBy keys=[[#0]]
                          Union
                            Negate
                              Project (#0)
                                Join on=(#0 = #1) type=differential
                                  ArrangeBy keys=[[#0]]
                                    Get l4
                                  ArrangeBy keys=[[#0]]
                                    Distinct group_by=[#0]
                                      Project (#0)
                                        Get l1
                            Get l4
                        Get l5
    With
      cte l5 =
        ArrangeBy keys=[[#0]]
          Project (#0, #2)
            ReadStorage mz_internal.mz_frontiers
      cte l4 =
        Distinct group_by=[#0]
          Project (#1)
            Get l3
  With Mutually Recursive
    cte l3 =
      Distinct group_by=[#0, #1]
        Union
          Get l2
          Project (#0, #3)
            Join on=(#1 = #2) type=differential
              ArrangeBy keys=[[#1]]
                Get l3
              ArrangeBy keys=[[#0]]
                Get l1
    cte l2 =
      Project (#0, #2)
        Join on=(#0 = #1) type=differential
          ArrangeBy keys=[[#0]]
            Get l0
          ArrangeBy keys=[[#0]]
            Get l1
    cte l1 =
      Union
        ReadIndex on=mz_compute_dependencies mz_compute_dependencies_ind=[*** full scan ***]
        Project (#0, #1)
          Join on=(#0 = #2) type=differential
            ArrangeBy keys=[[#0]]
              ReadIndex on=mz_object_dependencies mz_object_dependencies_ind=[differential join]
            ArrangeBy keys=[[#0]]
              ReadIndex on=mz_sinks mz_sinks_ind=[differential join]
    cte l0 =
      Union
        Project (#0)
          ReadIndex on=mz_indexes mz_indexes_ind=[*** full scan ***]
        Project (#0)
          ReadIndex on=mz_materialized_views mz_materialized_views_ind=[*** full scan ***]
        Project (#0)
          ReadIndex on=mz_sinks mz_sinks_ind=[*** full scan ***]

Used Indexes:
  - mz_internal.mz_indexes_ind (*** full scan ***)
  - mz_internal.mz_sinks_ind (*** full scan ***, differential join)
  - mz_internal.mz_materialized_views_ind (*** full scan ***)
  - mz_internal.mz_object_dependencies_ind (differential join)
  - mz_internal.mz_compute_dependencies_ind (*** full scan ***)
"
```

**Comment by teskje** [comment]
Thanks! I've added an index on `mz_frontiers` as that seems more generally useful. I verified locally that it makes the query fast, but lmk if you observe anything else!

**Comment by teskje** [comment]
@SangJunBak Not sure I follow! Can you provide an example of what can go wrong? The test in this PR uses a counter loadgen and things seem to work as expected.

**Comment by SangJunBak** [comment]
> Not sure I follow! Can you provide an example of what can go wrong? The test in this PR uses a counter loadgen and things seem to work as expected. 

This was a blunder on my part. Turns out I wasn't even using a counter source but rather the view was doing a cross join on `generate_series`.  Resolved offline with Jan too!

**Comment by teskje** [comment]
This is ready for review, I think!

File: test/testdrive/materialization-lag.td [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  WHERE
+    local_lag > INTERVAL '1s' OR
+    global_lag > INTERVAL '1s'
+
+> ALTER CLUSTER source SET (REPLICATION FACTOR 1)
+
+# Bring down the compute cluster and observe resulting lag.
+
+> ALTER CLUSTER compute SET (REPLICATION FACTOR 0)
+
```
Comment (jkosh44) [inline test/testdrive/materialization-lag.td:79]: Do we need a call to `mz_sleep` here to avoid flakiness?

**Review by jkosh44 — APPROVED** [review]

File: test/testdrive/materialization-lag.td (hunk 1 above)
Comment (umanwizard) [reply test/testdrive/materialization-lag.td:79]: Retries are considered table stakes for correctness in Testdrive IMO, I'm pretty sure we already have a huge amount of tests that will fail with retries turned off.

Comment (jkosh44) [reply test/testdrive/materialization-lag.td:79]: In that case, feel free to send it.

Comment (teskje) [reply test/testdrive/materialization-lag.td:79]: Yeah, most of the tests that query introspection relations include a blurb at the beginning of the .td file saying that this relies on testdrive's retries feature. I'll add that here too.

**Comment by teskje** [comment]
TFTR!

**Pull request merged**